### PR TITLE
CORE-963: Make BRCryptoWalletEvent a Reference Type

### PR DESCRIPTION
**Pull request by EBGToo**
Only converted BRCryptoWalletEvent at this time 1) for proof of concept and 2) to allow ADD_TRANSFER to be replaced with ADD_TRANSFERS in a future PR.

### CODE REVIEW

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c [hunk 1]
```diff
@@ -967,10 +963,9 @@ cryptoWalletManagerEstimateFeeBasisForPaymentProtocolRequest (BRCryptoWalletMana
                                                                    cookie,
                                                                    fee);
     if (NULL != feeBasis)
-        cryptoWalletGenerateEvent (wallet, (BRCryptoWalletEvent) {
-            CRYPTO_WALLET_EVENT_FEE_BASIS_ESTIMATED,
-            { .feeBasisEstimated = { CRYPTO_SUCCESS, cookie, feeBasis }} // feeBasis passed
-        });
+        cryptoWalletGenerateEvent (wallet, cryptoWalletEventCreateFeeBasisEstimated (CRYPTO_SUCCESS, cookie, feeBasis));
+
+    cryptoFeeBasisGive (feeBasis);
```
Comment (michael-brd) [inline WalletKitCore/src/crypto/BRCryptoWalletManager.c:968]: This give is outside the NULL check; is the NULL check actually useful or should this be under it?

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c [hunk 2]
```diff
@@ -945,10 +942,9 @@ cryptoWalletManagerEstimateFeeBasis (BRCryptoWalletManager manager,
                                                                      attributesCount,
                                                                      attributes);
     if (NULL != feeBasis)
-        cryptoWalletGenerateEvent (wallet, (BRCryptoWalletEvent) {
-            CRYPTO_WALLET_EVENT_FEE_BASIS_ESTIMATED,
-            { .feeBasisEstimated = { CRYPTO_SUCCESS, cookie, feeBasis }} // feeBasis passed
-        });
+        cryptoWalletGenerateEvent (wallet, cryptoWalletEventCreateFeeBasisEstimated (CRYPTO_SUCCESS, cookie, feeBasis));
+
+    cryptoFeeBasisGive (feeBasis);
```
Comment (michael-brd) [inline WalletKitCore/src/crypto/BRCryptoWalletManager.c:947]: This give is outside the NULL check; is the NULL check actually useful or should this be under it?

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c [hunk 3]
```diff
@@ -1093,10 +1088,7 @@ cryptoWalletManagerHandleTransferGENFilter (BRCryptoWalletManager cwm,
         cwm->listener.walletEventCallback (cwm->listener.context,
                                            cryptoWalletManagerTake (cwm),
                                            cryptoWalletTake (wallet),
-                                           (BRCryptoWalletEvent) {
-            CRYPTO_WALLET_EVENT_TRANSFER_ADDED,
-            { .transfer = { cryptoTransferTake (transfer) }}
-        });
+                                           cryptoWalletEventCreateTransfer(CRYPTO_WALLET_EVENT_TRANSFER_ADDED, cryptoTransferTake (transfer)));
```
Comment (michael-brd) [inline WalletKitCore/src/crypto/BRCryptoWalletManager.c:1091]: `cryptoWalletEventCreateTransfer` has a `cryptoTransferTake` inside it; is this one needed?

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c [hunk 4]
```diff
@@ -1106,10 +1098,7 @@ cryptoWalletManagerHandleTransferGENFilter (BRCryptoWalletManager cwm,
             cwm->listener.walletEventCallback (cwm->listener.context,
                                                cryptoWalletManagerTake (cwm),
                                                cryptoWalletTake (cwm->wallet),
-                                               (BRCryptoWalletEvent) {
-                                                CRYPTO_WALLET_EVENT_BALANCE_UPDATED,
-                                                { .balanceUpdated = { newBalance }}
-                                            });
+                                               cryptoWalletEventCreateBalanceUpdated(newBalance));
```
Comment (michael-brd) [inline WalletKitCore/src/crypto/BRCryptoWalletManager.c:1101]: Leak on `newBalance`? `cryptoWalletEventCreateBalanceUpdated` takes it, right?

File: WalletKitCore/src/crypto/BRCryptoWalletSweeper.c [hunk 5]
```diff
@@ -65,10 +67,9 @@ cryptoWalletManagerEstimateFeeBasisForWalletSweep (BRCryptoWalletSweeper sweeper
                                                                      fee);
     
     if (NULL != feeBasis)
-        cryptoWalletGenerateEvent (wallet, (BRCryptoWalletEvent) {
-            CRYPTO_WALLET_EVENT_FEE_BASIS_ESTIMATED,
-            { .feeBasisEstimated = { CRYPTO_SUCCESS, cookie, feeBasis }} // feeBasis passed
-        });
+        cryptoWalletGenerateEvent (wallet, cryptoWalletEventCreateFeeBasisEstimated (CRYPTO_SUCCESS, cookie, feeBasis));
+
+    cryptoFeeBasisGive(feeBasis);
```
Comment (michael-brd) [inline WalletKitCore/src/crypto/BRCryptoWalletSweeper.c:72]: This give is outside the NULL check; is the NULL check actually useful or should this be under it?

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java [hunk 6]
```diff
@@ -1322,7 +1324,7 @@ private static void handleWalletDeleted(Cookie context, BRCryptoWalletManager co
     }
 
     private static void handleWalletTransferAdded(Cookie context, BRCryptoWalletManager coreWalletManager, BRCryptoWallet coreWallet, BRCryptoWalletEvent event) {
-        BRCryptoTransfer coreTransfer = event.u.transfer;
+        BRCryptoTransfer coreTransfer = event.transfer();
```
Comment (michael-brd) [inline WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1327]: This isn't wrapped in a higher-level Java object (which IIRC does the `give()`), is it? Leak here?

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java [hunk 7]
```diff
@@ -1364,7 +1366,7 @@ private static void handleWalletTransferAdded(Cookie context, BRCryptoWalletMana
     }
 
     private static void handleWalletTransferChanged(Cookie context, BRCryptoWalletManager coreWalletManager, BRCryptoWallet coreWallet, BRCryptoWalletEvent event) {
-        BRCryptoTransfer coreTransfer = event.u.transfer;
+        BRCryptoTransfer coreTransfer = event.transfer();
```
Comment (michael-brd) [inline WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1369]: This isn't wrapped in a higher-level Java object (which IIRC does the `give()`), is it? Leak here?

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java [hunk 8]
```diff
@@ -1406,7 +1408,7 @@ private static void handleWalletTransferChanged(Cookie context, BRCryptoWalletMa
     }
 
     private static void handleWalletTransferSubmitted(Cookie context, BRCryptoWalletManager coreWalletManager, BRCryptoWallet coreWallet, BRCryptoWalletEvent event) {
-        BRCryptoTransfer coreTransfer = event.u.transfer;
+        BRCryptoTransfer coreTransfer = event.transferSubmit();
```
Comment (michael-brd) [inline WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1411]: This isn't wrapped in a higher-level Java object (which IIRC does the `give()`), is it? Leak here?

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java [hunk 9]
```diff
@@ -1448,7 +1450,7 @@ private static void handleWalletTransferSubmitted(Cookie context, BRCryptoWallet
     }
 
     private static void handleWalletTransferDeleted(Cookie context, BRCryptoWalletManager coreWalletManager, BRCryptoWallet coreWallet, BRCryptoWalletEvent event) {
-        BRCryptoTransfer coreTransfer = event.u.transfer;
+        BRCryptoTransfer coreTransfer = event.transfer();
```
Comment (michael-brd) [inline WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1453]: This isn't wrapped in a higher-level Java object (which IIRC does the `give()`), is it? Leak here?

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java [hunk 10]
```diff
@@ -1556,23 +1558,23 @@ private static void handleWalletFeeBasisUpdated(Cookie context, BRCryptoWalletMa
     }
 
     private static void handleWalletFeeBasisEstimated(Cookie context, BRCryptoWalletEvent event) {
-        BRCryptoStatus status = event.u.feeBasisEstimated.status();
+        BRCryptoWalletEvent.FeeBasisEstimate estimate = event.feeBasisEstimate();
```
Comment (michael-brd) [inline WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1561]: This isn't wrapped in a higher-level Java object (which IIRC does the `give()`), is it? Leak here?

**Review by michael-brd — COMMENTED** [review]
Looks really good. Some questions around gives/takes. I could be way off, just asking with an outsider's perspective at this point.

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c (hunk 1 above)
Comment (EBGToo) [reply WalletKitCore/src/crypto/BRCryptoWalletManager.c:968]: All the `give` function now accept a `NULL` argument.

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c (hunk 3 above)
Comment (EBGToo) [reply WalletKitCore/src/crypto/BRCryptoWalletManager.c:1091]: Ditto.

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c (hunk 2 above)
Comment (EBGToo) [reply WalletKitCore/src/crypto/BRCryptoWalletManager.c:947]: Ditto

File: WalletKitCore/src/crypto/BRCryptoWalletSweeper.c (hunk 5 above)
Comment (EBGToo) [reply WalletKitCore/src/crypto/BRCryptoWalletSweeper.c:72]: Ditto

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c (hunk 3 above)
Comment (EBGToo) [reply WalletKitCore/src/crypto/BRCryptoWalletManager.c:1091]: This is actually in a giant `#ifdef 0 ... #endif` block. (Hence I didn't find this when I scrubbed give/take use and hence I didn't see a leak when testing).  I'll fix it or remove the dead code.

File: WalletKitCore/src/crypto/BRCryptoWalletManager.c (hunk 4 above)
Comment (EBGToo) [reply WalletKitCore/src/crypto/BRCryptoWalletManager.c:1101]: Also in the `#ifdef 0 ... #endif` dead code block.

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java (hunk 6 above)
Comment (EBGToo) [reply WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1327]: Yes, probably leaked. I'll fix.

Comment (EBGToo) [reply WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1328]: The subsequent code, which didn't change, has a `try { ... } finally { coreTransfer.give(); }` so I think it is covered.  I turns out that `event` itself was not given and the `estimate` 'structure' from`BRCryptoWalletEvent.FeeBasisEstimate estimate = event.feeBasisEstimate();` didn't have its `basis` field given.  Fixed.

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java (hunk 7 above)
Comment (EBGToo) [reply WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1370]: Handled in subsequent `try ... finally`

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java (hunk 8 above)
Comment (EBGToo) [reply WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1412]: Handled in subsequent `try ... finally`

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java (hunk 9 above)
Comment (EBGToo) [reply WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1454]: Handled in subsequent `try ... finally`

File: WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java (hunk 10 above)
Comment (EBGToo) [reply WalletKitJava/corecrypto/src/main/java/com/breadwallet/corecrypto/System.java:1562]: The `estimate.basis` is either passed to `FeeBasis.create()` (on success) or explicitly given.

**Review by michael-brd — APPROVED** [review]

**Comment by EBGToo** [comment]
Note, commits given wrong JIRA.  Should be CORE-693

**Pull request merged**